Added more landing pads to the Airforce Command Center

Mailaender opened the pull request:
Closes https://github.com/OpenRA/ra2/issues/87.

sorcerer86pt commented:
Ok, tested this commit. 
Some changes need to be done. As you can see in this video:


[multiple airfields](https://drive.google.com/file/d/1wYVb4HsqdW45yZa2V3D_bBagN0hPF5IW/view?usp=sharing)

When multiple airfields exists, the airplanes sometimes enter in loop trying to find a slot, when another one is open (mostly when building new planes, when coming back from bombing mission, they find the slot correctly)

An possible fix is at build time, find a open slot in all airfields that the player controls and then deploy that plane there.

pgwipeout commented:
Your video is not accessible unfortunately.

sorcerer86pt commented:
Ok, should be gdrive permissions. I will fix that

A quarta, 24/06/2020, 16:20, Peter Geis <notifications@github.com> escreveu:

> Your video is not accessible unfortunately.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenRA/ra2/pull/735#issuecomment-648887157>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2N4ZYGJFGKAMBIUI73JXLRYIKVRANCNFSM4OA3NARQ>
> .
>

sorcerer86pt commented:
Try now. Fixed the permissions

>

Mailaender commented:
> An possible fix is at build time, find a open slot in all airfields that the player controls and then deploy that plane there.

Probably not possible with the way this hack works. Only 25% of the pads is constructing aircraft, the others are just for rearming.

yaaaaa commented:
May be it needs to spawn/build the plane right on free slot, and without flying/jumping after build?

sorcerer86pt commented:
I approve of that.

A quarta, 24/06/2020, 19:54, yaaaaa <notifications@github.com> escreveu:

> May be it needs to spawn/build the plane right on free slot, and without
> flying/jumping after build?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenRA/ra2/pull/735#issuecomment-649004190>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2N4Z4ERGQFPVY6BREHGB3RYJDXTANCNFSM4OA3NARQ>
> .
>

MustaphaTR commented:
Also from the video @sorcerer86pt sent looks like planes land facing 0, they should face 224. https://github.com/MustaphaTR/Romanovs-Vengeance/commit/cc69debf2138312526bcd7c59110544e2a44017b

Mailaender commented:
Changed the landing facing and added production capabilities to the pads. I don't know why `TakeOffOnCreation: false` is ignored. We have a ticket for that #683.

Phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,86 @@
+		public readonly HashSet<string> ActorTypes = new HashSet<string>();
+
+		[FieldLoader.Require]
+		[Desc("Locations to place the mines, from top-left of the building.")]
```
```suggestion
		[Desc("Locations to spawn the actors, relative to the top-left of the building.")]
```

Phrohdoh replied:
Is `self.Location` _guaranteed_ to be the top-left (and if so, for my edification, how/where)?
If not, I think we should say something like "relative to the origin of this actor."

Mailaender replied:
Look at the `Building` code or place an actor in the editor. The origin is always top-left. Removed the legacy cruft. This was used to spawn mines as in Generals China ability before and recycled.

Phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,86 @@
+		{
+			this.info = info;
+
+			SpawnActors(self);
```
Should this be invoked in `OnAddedToWorld` (or something similar), instead of in the trait ctor?
I don't actually know the answer, I'm asking to spark conversation or to learn why that isn't necessary.

Mailaender replied:
I think you are right and changed the code accordingly. This probably fixes nothing other than unnoticeable timing.

Phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,86 @@
+			actors.Clear();
+		}
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
```
I don't see this documented at all and this seems to be a very specific behavior unrelated to spawning nearby actors.
Regardless of whether this functionality remains, it should be documented IMO.

Mailaender replied:
Added a note to the trait description.

phrohdoh replied:
Is that in reference to the "which keep connected?"
That makes me think this'd be used for (visually) connecting walls, not ownership of the spawned actors always being that of the actor that spawned them throughout the game.

Reviewing this PR made me realize that the trait it introduces into RA2 seems like `FreeActor` with cell position offsets (instead of world) and selling, killing, and ownership changing of the free actors.
Would it make any sense to use (upgrading the implementation as necessary) multiple `FreeActor`s instead of a single similar `SpawnNeighboringActors`?

phrohdoh replied:
Owner changing is still undocumented.

Phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,86 @@
+			{
+				var cell = self.Location + offset;
+				var actor = info.ActorTypes.Random(self.World.SharedRandom).ToLowerInvariant();
+				{
```
Why are we creating another scope here?
Should this be a frame-end task?

Mailaender replied:
Yes, that was actually broken probably from an incomplete copy and paste. Fixed.

Phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,86 @@
+			}
+		}
+
+		public void RemoveActors()
```
Same here in general regarding undocumented behavior.

phrohdoh replied:
I've switched this comment back to unresolved since this still is not documented as far as I can tell.

Mailaender replied:
I explained "they keep connected" more explicitly now.

Phrohdoh commented on mods/ra2/rules/allied-structures.yaml:
```diff
@@ -349,6 +349,20 @@ gaairc:
+		ActorTypes: gapad
+		Locations: 1,0, 1,1, 2,1
+
+gapad:
```
Why, technically, are we splitting this into multiple actor types?
Is this required to work around an engine issue?
If not, to me this seems like (prematurely) building more than we need.

Mailaender replied:
There can't be more than one docking activity per actor. This is an engine limitation which can't be resolved here.

Phrohdoh commented:
`SpawnNeighboringActors` from a modding perspective (given the information this PR as-is makes available to modders) _seems_ awfully similar to `FreeActor`, so I believe differentiating via docs/name would be beneficial.

Mailaender commented:
Updated and fixed an issue where selling the aircraft command would not remove the additional pads.

phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,97 @@
+	{
+		[FieldLoader.Require]
+		[ActorReference]
+		[Desc("Types of actors to place. If multiple are defined, a random one will be selected.")]
```
```suggestion
		[Desc("Types of actors to place. If multiple are defined, a random one will be selected for each actor spawned.")]
```

To make it more clear that each spawned actor will be a random type from this set, instead of a random single type being selected for all to-be-spawned actors.

Mailaender replied:
Updated.

phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,97 @@
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			foreach (var actor in actors)
```
Do we need to check for exists/alive/in-world here (or does owner change machinery handle that for us?) and/or keep this list up-to-date (removing from the list when referenced actors die or enter/exit world)?

This trait is clearly intended to be used for spawning buildings (the aircraft pads in specific), so would it make sense to bake that intention into the code (no need to be overly generic IMO, especially if we aren't going to consider all of the possible interactions and events we need to think about)?

Mailaender replied:
I think not as this trait is really used for these invisible invincible fake actors.

phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,97 @@
+				actor.ChangeOwnerSync(newOwner);
+		}
+
+		void INotifyKilled.Killed(Actor self, AttackInfo e)
```
Undocumented behavior.

Mailaender replied:
Well kinda.

phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,97 @@
+		}
+
+		void INotifySold.Selling(Actor self) { }
+		void INotifySold.Sold(Actor self)
```
Undocumented behavior.

phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,97 @@
+	public class SpawnNeighboringActors : INotifyKilled, INotifyOwnerChanged, INotifyActorDisposing, INotifyAddedToWorld, INotifySold
+	{
+		readonly SpawnNeighboringActorsInfo info;
+		readonly List<Actor> actors = new List<Actor>();
```
I don't see items being added to this list (which could be a hashset? not sure if that'll really matter).
Given that, I don't understand how this trait actually keeps ownership/existence in sync.

Mailaender replied:
That is a bug. Fixed below.

phrohdoh commented on OpenRA.Mods.RA2/Traits/SpawnNeighboringActors.cs:
```diff
@@ -0,0 +1,97 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
Not entirely related to this PR in specific, but should copyright of non-engine/core code be attributed to engine/core?  Also copy+pasting `2007-<year>` around doesn't actually make that the case.

Mailaender replied:
I won't change just this file for consistency.

phrohdoh commented on mods/ra2/rules/allied-structures.yaml:
```diff
@@ -349,6 +349,20 @@ gaairc:
 		Actors: e1, e1
+	SpawnNeighboringActors:
+		ActorTypes: gapad
+		Locations: 1,0, 1,1, 2,1
```
Why only 3?
This seems to line up, in some way I don't yet understand, with the footprint having 3 `=`s, not 4 (or even 6 `x`s).

Mailaender replied:
The main building is already one pad.

phrohdoh commented on mods/ra2/rules/allied-structures.yaml:
```diff
@@ -349,6 +349,20 @@ gaairc:
+	Building:
+		Footprint: x
+		Dimensions: 1,1
+		AllowInvalidPlacement: true
```
Why?

Mailaender replied:
So it can be placed on top of the main building.

phrohdoh commented on mods/ra2/rules/allied-structures.yaml:
```diff
@@ -283,7 +283,7 @@ gaairc:
 	ProvidesPrerequisite@default:
 	Building:
-		Footprint: xxx xxx
+		Footprint: x=x x==
```
Why this footprint in particular?

Mailaender replied:
The first `x´ is the pad that comes with the main building. The 3 `=` are needed so another actor can be placed on top of it providing 1 more pad each.

phrohdoh reviewed (changes requested):
Inline comments regarding undocumented behavior, making the trait only as abstract/generic as necessary (this isn't engine/core where we have and expect 3rd-party reuse (and if that happens, things can be made more generic then)), how the trait logic works in general (re not seeing items added to the `actors` instance member), and some MiniYaml values.

Mailaender commented:
Rebased and updated.

MustaphaTR commented:
Needs rebase.

abcdefg30 commented:
Tbh I'm not sure if this PR is necessary. If this repository seems unmaintained to people, I don't think there is a reason to add this workaround, or is there? I.e. since we're not in a rush we can wait for docking to work like we need.
